### DISCUSSION TITLE
在Safari里安装脚本，油猴和ADGuard怎么选择？

### DISCUSSION THREAD

**Discussion by LoneFireBlossom** [the1812/Bilibili-Evolved#2362] [综合讨论 / General]
看到Bilibili-Evolved显示在ADGuard里功能不全，但是在油猴里功能完整。
而有的脚本安装在油猴里就无效（比如 https://greasyfork.org/zh-CN/scripts/372064-知乎重排for印象笔记 这个），安装在ADGuard里就有效。
再比如这个帖子里的讨论： https://v2ex.com/t/762679

进而想到会不会安装在油猴和ADGuard上脚本的运行速度也会有显著差别。因为目前安装在油猴里的脚本，就比如这个bilibili-Evolved，经常并不是我打开B站就有，而是得过一会儿才会出现功能按钮。Chrome就没有这个问题。（可是Chrome也太尼玛耗电+发热了，看个B站1080p视频2倍速，风扇直接起飞满速转……）

搜了一下网上关于这方面的讨论没找到什么有用的，感觉似乎只能一个一个脚本，在两边分别尝试安装了吧……

**Comment by the1812** [comment]
这得看脚本是对着哪个写的了, 所以只能一个个试

**Comment by LoneFireBlossom** [comment]
干脆在这里记录一下得了。
下面这个「片假名终结者」脚本在Safari油猴里无效，在AdGuard里有效。
（注意在AdGuard里装了就得把油猴里的脚本关掉）
https://github.com/Arnie97/katakana-terminator

**Comment by LoneFireBlossom** [comment]
popup search 和 select like opera 如果安装在Safari油猴里，有些网站会不支持，比如GitHub和采用类似Devonthink和美卡论坛那种论坛模板的网站。装在AdGuard里就支持上面两种情况了。不知道是不是适配所有网站。
https://github.com/lkytal/GM/raw/master/popsearch.user.js
https://git.oschina.net/coldfire/GM/raw/master/select.user.js

**Reply by LoneFireBlossom** [reply]
popup search装在AdGuard里会造成某些网页点击链接时出现问题，但是装在Safari油猴里没问题。
目前测试有问题的网页有：
- surge的反馈论坛 https://community.nssurge.com ，点击帖子不会进入帖子的链接，得刷新。后退也有问题。

于是设置了
// @exclude                 *community.nssurge.com/*

**Comment by LoneFireBlossom** [comment]
这俩脚本也是在楼上所说的一些网站里没用，装在Adguard里就好了。
不过这俩脚本的设置窗口都是需要在油猴的那个按钮里才能点开，装在AdGuard里我就不知道怎么点了……我点进去看发现以我这只上过本科基础C语言课的能力根本改不了触发的快捷键，只能是把第一个脚本自动显示大图的时间调大了一下（搜索delay我改成了10000）
https://greasyfork.org/scripts/394820-mouseover-popup-image-viewer/code/Mouseover%20Popup%20Image%20Viewer.user.js
https://greasyfork.org/scripts/2998-search-by-image/code/Search%20By%20Image.user.js

**Comment by LoneFireBlossom** [comment]
网页限制解除也是。
https://greasyfork.org/scripts/28497-remove-web-limits-modified/code/Remove%20web%20limits(modified).user.js

**Comment by LoneFireBlossom** [comment]
关于Safari油猴某些脚本不能生效的原因解释
https://github.com/XIU2/UserScript/issues/107

**Comment by LoneFireBlossom** [comment]
这几个也是装在AdGuard里就好了

https://greasyfork.org/zh-CN/scripts/24508-userscript-show-site-all-userjs
https://greasyfork.org/zh-CN/scripts/412245-github-增强-高速下载
https://greasyfork.org/zh-CN/scripts/429714-新标签页打开链接

**Comment by LoneFireBlossom** [comment]
这个脚本安装在AdGuard里就可以在知乎等搜索页面里生效了。安装在Safari油猴里的话在知乎搜索页面是没有的
https://greasyfork.org/zh-CN/scripts/27752-searchenginejump-搜索引擎快捷跳转

试了一下安装在AdGuard里，在GitHub上也是无效的，明明代码里有
```
// @match          *://*.github.com/*
```
但是在Chrome里也无效啊，不懂这个match有什么意义

**Comment by LoneFireBlossom** [comment]
这个装在油猴和AdGuard里对Safari都无效，测试了Google、百度贴吧、discuz!论坛、3dm
在Chrome油猴里是有效的。

https://greasyfork.org/zh-CN/scripts/419215-自动无缝翻页

---

这个脚本也是如此，其中的自动翻页功能装在哪一边对Safari都无效。
https://greasyfork.org/zh-CN/scripts/424246-v2ex-增强

---

这个脚本的自动翻页有效，装在Safari油猴里。
https://greasyfork.org/en/scripts/412890-3dm论坛增强
https://greasyfork.org/zh-CN/scripts/414005-全球主机交流论坛增强
https://greasyfork.org/zh-CN/scripts/412680-吾爱破解论坛增强-自动签到-翻页

---

AC-Baidu的自动翻页也是无效的。
https://greasyfork.org/zh-CN/scripts/14178-ac-baidu-重定向优化百度搜狗谷歌必应搜索-favicon-双列

---

但是endless Google在Google上是有效的，装在Safari油猴里即可。我猜想可能两者实现方式不一样（没看源码）
https://openuserjs.org/scripts/tumpio/Endless_Google

**Reply by 2henVVen** [reply]
请问如何选择将脚本装在油猴还是adguard中? 在安装了adguard后在点击安装此脚本后每次都将我的脚本装在adguard中, 而不是油猴.

**Reply by LoneFireBlossom** [reply]
> 请问如何选择将脚本装在油猴还是adguard中? 在安装了adguard后在点击安装此脚本后每次都将我的脚本装在adguard中, 而不是油猴.

右键安装链接，复制一下就能看到为啥会在adguard里打开了。
在没点过安装链接的情况下，右键安装链接，在新标签页中打开

**Reply by 2henVVen** [reply]
谢谢回复.但还是没看明白, 
在点过安装链接后, 怎么办? 在新标签页中打开还是跳转到adguard

**Reply by 2henVVen** [reply]
> > 请问如何选择将脚本装在油猴还是adguard中? 在安装了adguard后在点击安装此脚本后每次都将我的脚本装在adguard中, 而不是油猴.
> 
> 右键安装链接，复制一下就能看到为啥会在adguard里打开了。 在没点过安装链接的情况下，右键安装链接，在新标签页中打开

adguard:userscript?location=https://update.greasyfork.org/scripts/419224/蓝奏云网盘增强.user.js
好像懂了, 直接把前面删除就好了

**Comment by wyq977** [comment]
AdGuard, Macaque, 油猴, Stay 2, Userscripts, MeddleMonkey 这几个你用的怎么样啊？@LoneFireBlossom 我有一个script只能AdGuard用

**Reply by wyq977** [reply]
https://greasyfork.org/zh-CN/scripts/438841-复制成-markdown-的-link
